mlx/ops.cpp

```diff
@@ -1330,6 +1330,45 @@ array where(
       inputs);
 }
 
+array nan_to_num(
+    const array& a,
+    const float& nan,
+    const float& posinf,
+    const float& neginf,
+    StreamOrDevice s /* = {} */) {
+  Dtype dtype = a.dtype();
+
+  if (!issubdtype(dtype, inexact)) {
+    return a;
+  }
+
+  nan_ = array(nan, dtype);
+  posinf_ = array(posinf, dtype);
+  neginf_ = array(neginf, dtype);
+
+  // How to find the max value of a the same dtype as a?
+
+  // Checking if input values are too large for float16 and bfloat16
+  if (nan_ == std::numeric_limits<dtype>::infinity() ||
+      nan_ == -std::numeric_limits<dtype>::infinity()) {
+    nan_ = array(0.0, dtype);
+  }
+  if (posinf_ == std::numeric_limits<dtype>::infinity() ||
+      posinf_ == -std::numeric_limits<dtype>::infinity()) {
+    posinf_ = array(std::numeric_limits<dtype>::max());
+  }
+  if (neginf_ == std::numeric_limits<dtype>::infinity() ||
+      neginf_ == -std::numeric_limits<dtype>::infinity()) {
+    neginf_ = array(-std::numeric_limits<dtype>::max());
+  }
```

Comment:
Instead of this, I would:

- Make the inputs limits optional (e.g. `const std::optional<float>& nan` 
- check if the input has a value. Use it if so.
- If it doesn't have a value set it correctly kind of like how you are already doing. But you'll need to have a branch because `numeric_limits<dtype>` is not valid.

So something like:
```c++
if (dtype == float32) {
 posinf = std::numeric_limits<float>::max();
} else if (dtype == float16) {
...
} else if (dtype == bfloat16) {
...
}
```

Reply:
Thank you for your comments! Everything is looking good apart from the issue below.
Currently part of the branch looks like this:
```cpp
if (posinf.has_value()) {
    posinf_ = posinf.value();
  } else if (dtype == float32) {
    posinf_ = std::numeric_limits<float>::max();
  } else if (dtype == float16) {
    posinf_ = std::numeric_limits<float16_t>::max();
  } else if (dtype == bfloat16) {
    posinf_ = std::numeric_limits<bfloat16_t>::max();
  }
  ```
It compiles fine, but when tested:
- `std::numeric_limits<float16_t>::max();` returns `0` (or `0.0F` to be precise). (In fact `min()` and `infinity()` all do for `float16_t`. Is this intended?)
-  `std::numeric_limits<bfloat16_t>::max();` throws `RuntimeError: Item size 2 for PEP 3118 buffer format string B does not match the dtype B item size 1.`